File: java/QuickSort.java
public class QuickSort {
    // A utility function to swap two elements
    static <T> void swap(T[] arr, int i, int j) {
        T temp = arr[i];
        arr[i] = arr[j];
        arr[j] = temp;
    }

    // This function takes last element as pivot,
    // places the pivot element at its correct position
    // in sorted array, and places all smaller to left
    // of pivot and all greater elements to right of pivot
    static <T extends Comparable<T>> int partition(T[] arr, int low, int high) {
        // Choosing the pivot
        T pivot = arr[high];

        // Index of smaller element and indicates
        // the rightposition of pivot found so far
        int i = (low - 1);

        for (int j = low; j <= high - 1; j++) {
            // If current element is smaller than the pivot
            if(arr[j].compareTo(pivot) < 0) {
                // Increment index of smaller element
                i++;
                swap(arr, i, j);
            }
        }
        swap(arr, i + 1, high);
        return (i + 1);
    }

    // The main function that implements QuickSort
    // arr[] --> Array to be sorted,
    // low --> Starting index,
    // high --> Ending index
    static <T extends Comparable<T>> void quickSort(T[] arr, int low, int high) {
        if (low < high) {
            // pi is partitioning index, arr[p]
            // is now at right place
            int pi = partition(arr, low, high);

            // Separately sort elements before
            // partition and after partition
            quickSort(arr, low, pi - 1);
            quickSort(arr, pi + 1, high);
        }
    }

    // To print sorted array
    public static <T> void printArr(T[] arr) {
        for (T element : arr)
            System.out.print(element + " ");
        System.out.println();
    }
}

File: src/quicksort.rs
pub enum Scheme {
    Lomuto,
    Hoare,
}

pub fn quicksort<T: PartialOrd + Copy>(arr: &mut [T], scheme: Scheme) {
    _quicksort(arr, 0, (arr.len() - 1) as isize, &scheme);
}

fn _quicksort<T: PartialOrd + Copy>(arr: &mut [T], left: isize, right: isize, scheme: &Scheme) {
    if left <= right {
        let partition_idx = match scheme {
            Scheme::Lomuto => lomuto(arr, left, right),
            Scheme::Hoare  => hoare(arr, left as usize, right as usize),
        };

        _quicksort(arr, left, partition_idx - 1, &scheme);
        _quicksort(arr, partition_idx + 1, right, &scheme);
    }
}

fn hoare<T: PartialOrd + Copy>(arr: &mut [T], left: usize, right: usize) -> isize {
    let pivot = arr[left];
    let mut i = left;
    let mut j = right;

    loop {
        while arr[i] < pivot {
            i += 1;
        }

        while arr[j] > pivot {
            j -= 1;
        }

        if i >= j {
            return j as isize;
        }

        arr.swap(i, j);
        i += 1;
        j -= 1;
    }
}

fn lomuto<T: PartialOrd + Copy>(arr: &mut [T], left: isize, right: isize) -> isize {
    let pivot = arr[right as usize];
    let mut i = left;

    for j in left..right {
        if arr[j as usize] <= pivot {
            arr.swap(i as usize, j as usize);
            i += 1;
        }
    }

    arr.swap(i as usize, right as usize);

    i
}